### PR TITLE
Implement workshop booking

### PR DESCRIPTION
**Pull request by marcoacierno**

### CODE REVIEW

File: backend/api/schedule/mutations.py [hunk 1]
```diff
@@ -42,8 +45,70 @@ class UpdateOrCreateSlotItemResult:
...
+        self, info, id: strawberry.ID, conference: str
+    ) -> typing.Union[ScheduleItemBookingStatus, ReleaseScheduleItemBookingError]:
+        user = info.context["request"].user
+
+        schedule_item_booking = ScheduleItemBooking.objects.filter(
+            user=user, schedule_item_id=id, schedule_item__conference__code=conference
+        ).first()
+
+        if not schedule_item_booking:
+            return ReleaseScheduleItemBookingError(message=_("No booking found"))
```
Comment (patrick91) [inline backend/api/schedule/mutations.py]: this could a BookingNotFound error ;)

Comment (patrick91) [reply backend/api/schedule/mutations.py]: same for the other ones, I imagine the FE using the typename information to show an appropriate message 😊

File: backend/api/schedule/types.py [hunk 2]
```diff
@@ -35,6 +37,25 @@ class ScheduleItem:
...
+    @strawberry.field
+    def is_booked(self, info) -> bool:
+        user = info.context["request"].user
+
+        if not user.is_authenticated:
+            return False
+
+        return ScheduleItemBooking.objects.filter(
+            user=user, schedule_item=self
+        ).exists()
```
Comment (patrick91) [inline backend/api/schedule/types.py]: we need a way to optimise this at some point 😊

File: backend/api/schedule/types.py [hunk 3]
```diff
@@ -35,6 +37,25 @@ class ScheduleItem:
...
+
+    @strawberry.field
+    def can_book(self, info) -> bool:
+        return self.allows_booking and self.capacity_left > 0
```
Comment (patrick91) [inline backend/api/schedule/types.py]: we could reuse the resolver 😊

```suggestion
    def can_book(self, info) -> bool:
        return self.allows_booking and self.has_free_spot(info)
```

Comment (patrick91) [reply backend/api/schedule/types.py]: we could also move the resolver to be a property on the model, but that's minor

File: backend/schedule/models.py [hunk 4]
```diff
@@ -131,6 +133,21 @@ class ScheduleItem(TimeStampedModel):
         on_delete=models.PROTECT,
     )
 
+    maximum_capacity = models.PositiveIntegerField(
+        _("maximum capacity"), blank=True, null=True
+    )
+
+    @property
+    def allows_booking(self):
+        return self.type == ScheduleItem.TYPES.training
```
Comment (patrick91) [inline backend/schedule/models.py]: we need to take into account also SUBMISSION types where the submission is a training :)

File: backend/schedule/models.py [hunk 5]
```diff
@@ -131,6 +133,21 @@ class ScheduleItem(TimeStampedModel):
...
+
+    @property
```
Comment (patrick91) [inline backend/schedule/models.py]: let's make this a cached property :)

Comment (marcoacierno) [reply backend/schedule/models.py]: done

File: backend/api/schedule/types.py (hunk 3 above)
Comment (marcoacierno) [reply backend/api/schedule/types.py]: moved it to the model

**Pull request closed without merging**